### PR TITLE
Latin hyphenation pattern

### PR DESCRIPTION
**Pull request by jperon**
I hope I didn't forget anything.

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/crengine/393)

### CODE REVIEW

File: crengine/src/textlang.cpp [hunk 1]
```diff
@@ -367,6 +368,7 @@ static quotes_spec _quotes_spec_table[] = {
     { "ln",       U"\x201c", U"\x201d", U"\x2018", U"\x2019" }, /* “ ” ‘ ’ */
     { "lo",       U"\x201c", U"\x201d", U"\x2018", U"\x2019" }, /* “ ” ‘ ’ */
     { "lrc",      U"\x201c", U"\x201d", U"\x2018", U"\x2019" }, /* “ ” ‘ ’ */
+    { "fr",       U"\x00ab", U"\x00bb", U"\x201c", U"\x201d" }, /* « » “ ” */  /* "la" just as "it" */
```
Comment (poire-z) [inline crengine/src/textlang.cpp:371]: You forgot to change `"fr"`.

Comment (NiLuJe) [reply crengine/src/textlang.cpp:371]: It's also not in alphabetical order at that point in the list ;).

**Comment by poire-z** [comment]
Pinging @roshavagarga for a quick review.

**Comment by roshavagarga** [comment]
@poire-z @jperon Looks fairly good to me. I can't take a look at the hyph file itself, so please make sure it's formatted like this one:
https://github.com/koreader/crengine/blob/master/cr3gui/data/hyph/Armenian.pattern, specifically the beginning and link to source file.

I'm guessing you used this source: https://github.com/hyphenation/tex-hyphen/blob/master/hyph-utf8/tex/generic/hyph-utf8/patterns/tex/hyph-la.tex

There are also Classic Latin and Liturgical Latin, though I have no idea what the difference between Latin and Classic Latin is :)

**Comment by roshavagarga** [comment]
Okay, I now saw that this is actually based on a separate project. Whether the stuff in http://www.hyphenation.org/tex#languages would be useful is beyond me, but I'd like some more info:
1. Is this purely for Liturgical Latin?
2. If so, should it be renamed in some way, in case there are important distinctions? (I'm guessing there are, since hyph.org has 3 separate types of Latin)

**Comment by jperon** [comment]
Good point. So here is what I suggest (last commit):

- as hyph.la.modern (which is the base of hyphen-la.tex) [isn't maintained anymore](https://github.com/gregorio-project/hyphen-la/tree/6032c24798f71287d51e67dabe2bdaecb82d2490/patterns/generation), and differences with hyph.la.liturgical (which is the base of hyph-la-x-liturgic.tex) isn't that big, use [hyph-la-x-liturgic.tex](https://github.com/hyphenation/tex-hyphen/blob/master/hyph-utf8/tex/generic/hyph-utf8/patterns/tex/hyph-la-x-liturgic.tex) as Latin.pattern;
- add Lattin-classic.pattern, based on [hyph-la-classic.tex](https://github.com/hyphenation/tex-hyphen/blob/master/hyph-utf8/tex/generic/hyph-utf8/patterns/tex/hyph-la-x-classic.tex).

As of differences between classical and liturgical Latin, they're negligible about grammar, but quite important about pronunciation, hence hyphenation differences.

**Comment by poire-z** [comment]
```
$ .ci/script.sh
Warning, json should reflect hyphenation patterns. Diff:
24d23
< Latin-classic.pattern
```
You need to add an entry in language.json for the other pattern (we don't use this file anymore, it's just there as a remnant, as some kind of duplicate reference).

It seems there is a unique lang tag for latin: "la". So both classic latin and liturgical latin books will have "la" as their lang. So, with this, liturgical (modern?) will be the default.
(We currently can't allow selecting from multiple hyph dicts per a single lang tag... might be quite a bit of work to allow that, so let's forget it.)

I would think classical latin ought to be the default (as scholars reading the old classics would expect - as there's no modern crime novels in latin :), but I know nothing about that :)
When reading pre-18th philosophy (St Augustin, Thomas d'Acquin, Descartes, Spinoza...) in their original latin text, what hyphenation would be prefered ? modern/liturginal or classic ?

(Anyway, as the non-classic pattern is way smaller than the classic one - and as you're the one introducing hyphenation for latin, I'm fine with you deciding which is to be the default if there's no clear cut.)

**Comment by Frenzie** [comment]
> You need to add an entry in language.json for the other pattern (we don't use this file anymore, it's just there as a remnant, as some kind of duplicate reference).

Sounds like both the file and the check can be tossed then? :-)

> When reading pre-18th philosophy (St Augustin, Thomas d'Acquin, Descartes, Spinoza...) in their original latin text, what hyphenation would be prefered ? modern/liturginal or classic ?

I'd say that depends. Aquinas strikes me as a liturgical type, but Erasmus is more into retro-classical. Church Latin is pronounced the way it is because it's basically written Latin with what since roughly the 14th century we've called Italian (Tuscan) sounds. But iirc until Erasmus reconstructed the classical pronunciation, it roughly went along with the Latinate vernaculars.

**Comment by jperon** [comment]
> > You need to add an entry in language.json for the other pattern (we don't use this file anymore, it's just there as a remnant, as some kind of duplicate reference).
> 
> Sounds like both the file and the check can be tossed then? :-)

Or reverse: hardcoded values put into this file so that it's easier to add personal hyphen dictionaries ? (I wouldn't dare asking for anything though, as my C++ skills are less than nothing).

**Comment by jperon** [comment]
@poire-z For sure I'm a bit biased on that point (being a priest)… But I think that putting the liturgical form as a default would be more coherent with the behaviour of (La)TeX, as it is the nearest to `hyphen-la`. In any case, typographical rules aren't so far away in the menu: should anyone be surprised by default choice, changing it wouldn't be so complicated.

**Comment by Frenzie** [comment]
> Or reverse: hardcoded values put into this file so that it's easier to add personal hyphen dictionaries ? (I wouldn't dare asking for anything though, as my C++ skills are less than nothing).

I think that's more or less how it used to be. I added the check a couple of years ago because people would add a hyphenation pattern and then it wouldn't show up in KOReader. @poire-z will recall the further details and reasoning better.

**Comment by poire-z** [comment]
I added the textlang stuff mostly to get better line breaking per language, and to get that per-local HTML lang tags. So, if felt logical hyphenation should follow that, and become local (so, no more global stuff driven  by a hyphenation selection from frontend).
(We then get a single "typography language", that drives all the things that depends on the language, and it assumes it's "constant" per language.)
If then felt easier to hardcode all that in crengine - easier, just to avoid parsing some config file, and to avoid having to drive all that by the frontend, which is what I had in mind first, but given all the little rules, if felt easier to hardcode that and have all the logic in one place: textlang.cpp. As a side effect, it makes the integration easier for other frontends like CoolReader.
Otherwise, we'd have to have all the default quotes and line-breaking language tweaks driven in that json file, and parse it, and pilot the code from that. Not keen on going that way :) 
So yes, sad side effect: not easy to add user hyphenation, except by overwritting an existing hyph.pattern file with one's own.

> But I think that putting the liturgical form as a default would be more coherent with the behaviour of (La)TeX, as it is the nearest to hyphen-la.

OK, but so, with these philosophy text I mentionned, what one would be using with Tex: la(modern) or la-classic ?
(If the answer is la(modern), and la-liturgical is the most similar to la(modern), then fine. Otherwise, less fine :)

> In any case, typographical rules aren't so far away in the menu: should anyone be surprised by default choice, changing it wouldn't be so complicated.

Yes, but the one you set as the default for "la" will be used for all latin books (and embedded latin fragments with `lang="la"` in other language books).
Unless a user set la-cla as its DEFAULT typography language (so, applying even on english or french text), he will get your "la" default, and he will have to go change it in that menu to "Latin (classic)" for all new lang="la" books he opens :(

**Comment by Frenzie** [comment]
Could you perhaps give a few concrete examples of differences?

**Comment by jperon** [comment]
> OK, but so, with these philosophy text I mentionned, what one would be using with Tex: la(modern) or la-classic ?
> (If the answer is la(modern), and la-liturgical is the most similar to la(modern), then fine. Otherwise, less fine :)

For philosophy text from authors you mentionned, I think `hyphen-la` (= modern, not maintained) would be the default choice. `la-classic` would be used mostly for older texts (Cicero etc.). About LaTeX, if one uses `la-classic`, that means he's really aware of the distinction.

**Comment by jperon** [comment]
@Frenzie The best examples are on [hypen-la](https://github.com/gregorio-project/hyphen-la/tree/master/doc#comparison-of-the-three-hyphenation-styles).

**Comment by Frenzie** [comment]
I have suddenly acquired a very strong preference for classical.

> For philosophy text from authors you mentionned, I think hyphen-la (= modern, not maintained) would be the default choice. la-classic would be used mostly for older texts (Cicero etc.).

I think that needs more support, seeing how Cicero/Seneca/Ovid is was seen as the example to follow for the likes of Descartes and Spinoza. In other words, it seems to me that your example argues the exact opposite.

**Comment by jperon** [comment]
As I said, the question doesn't matter so much for me; if your preference goes in that direction, let's go for "Latin.pattern" / "Latin_liturgic.pattern". Just I wouldn't spend hours on such a small problem…

**Comment by poire-z** [comment]
> Just I wouldn't spend hours on such a small problem…

Well, if you go with default to classic, it's _you_ who will have to select "Latin (Liturgic") for each of you latin book :/ 
May be we/I will have to do something to allow more customisation per language, if HTML and RFCs only provide "la" for latin, but there's more variations for typography/hyphenation.

(It seems ancient greek is "grc", vs modern greek is "el", so they are distinguished.)

**Comment by roshavagarga** [comment]
@poire-z Sidenote, our Greek is probably the wrong type, 3 options for that too in hyph.org :)

**Comment by jperon** [comment]
> > Just I wouldn't spend hours on such a small problem…
> 
> Well, if you go with default to classic, it's _you_ who will have to select "Latin (Liturgic") for each of you latin book :/
> May be we/I will have to do something to allow more customisation per language, if HTML and RFCs only provide "la" for latin, but there's more variations for typography/hyphenation.

Sure, but I don't care so much: I mostly read quite long books, so I won't have to set it too often! If you find time to enhance it, many thanks; but you certainly have greater priorities!

**Comment by poire-z** [comment]
Current PR is missing Latin_liturgical.pattern.

> The best examples are on [hypen-la](https://github.com/gregorio-project/hyphen-la/tree/master/doc#comparison-of-the-three-hyphenation-styles).

@roshavagarga : any third opinion on which latin hyph should be the default ?

**Comment by roshavagarga** [comment]
I really don't have an opinion one way or the other, since my knowledge of Latin is confined to a few pig Latin phrases used by Pratchett :)

**Comment by jperon** [comment]
@poire-z Honestly, don't care too much about it: from an "ecclesiastical" point of view, the only disturbing case in the examples listed would be `mag - nus` (as `gn` is one phoneme, like spanish `ñ`): other cases wouldn't even be noticed by an uninformed person (they wouldn't be "grammatically incorrect"). This would be far better, in any case, than algorithmic or English hyphenation!

**Comment by poire-z** [comment]
Thanks for giving in :) but I just want Descartes to be as well served as Cicero, and as I lack any knowledge in the area of latin, I'm just asking so we're all sure - and it's not just a matter of personnal preferences (of course, if I had a preference, I would fight too :)
So, if I get my lesson from all your points:
- Cicero and his buddies would have liked la-classic
- Thomas and later used to be read with italian phonetics (whatever italian might have been at the time) - is there a specific time/period for this shift?
- Then Erasmus came in and teach everybody how latin was pronounced way back in the past, and that everybody should do it that way now
- But the Church didn't listen, and has been keeping italian phonetics
- But everybody else (scholars and free thinkers) followed erasmus, and it's what's expected even today in universities, as it helps with etymology (and that's why you're fine with providing la-classic?)

(Right ? :)

**Comment by jperon** [comment]
I think you got it to a point it isn't even caricatural… Though we could add that, for Descartes, he probably used another pronunciation, that got used in France till (roughly) the beginning of 20th century, and that was basically French pronunciation (hence "Té Déom" (if I use French writing) instead of "Té Deoum"). So, as you see, we make it even simpler than it is…

What I would add is that, IMO, we shouldn't care about how Descartes himself or Cicero would have hyphenated (about that, we know even less, I think, than about their pronunciation…), but about the expectation of those who, today, read their books. That's where I say that, IMHO (but perhaps I'm biased, as I certainly know more Catholic priests than many others…), most readers of philosophical books from (roughly) S. Augustine to (roughly) the 18th - 19th century would be surprised by `mag-nus`, while it would be the opposite for "classical" readers or today universities.

But, honestly, I don't think anyone of them would complain.

**Comment by jperon** [comment]
>     * Thomas and later used to be read with italian phonetics (whatever italian might have been at the time) - is there a specific time/period for this shift?

Just realize I didn't see this part of the question: I didn't study that historical point myself, but according to Hans Ørberg in the *Student's Manual* of his method *Lingua Latina Per Se Illustrata*:

> *Classical pronuciation described above was that of educated Romans in the first century B.C. In imperial times (1st−5th centuries A.D.) the pronunciation of Latin underwent considerable changes […]. The main features of this Late Latin pronunciation survive in the pronunciation of Latin still used in Italy. This 'Italian' pronunciation of Latin is widely used in the Roman Catholic Church and in church singing.*

**Comment by Frenzie** [comment]
@poire-z 
> > Thomas and later used to be read with italian phonetics (whatever italian might have been at the time) - is there a specific time/period for this shift?

The 14th century is when they standardized it as "Italian," but the pronunciation was a series of changes over the previous millennium, some of which peevers already complained about around 200.

@jperon 
> That's where I say that, IMHO (but perhaps I'm biased, as I certainly know more Catholic priests than many others…), most readers of philosophical/theological books from (roughly) S. Augustine to (roughly) the 18th - 19th century would be surprised by mag-nus

There may be another factor there besides the Catholic priests, along the lines of "in Italy and France." In the low countries and the British Isles Latin grammar and pronunciation were perhaps somewhat paradoxically much less affected by our native tongues because it's always been a completely foreign language.

Anyway, I randomly looked up Erasmus' most famous text, clicked on a random page [here](https://archive.org/details/moriaeencomiumid00eras/page/312/mode/2up), and it didn't disappoint. (Of course this is a sample size of 1, so take it with a giant grain of salt.)

Edition of Moriae encomium from 1540 (ig-noscitur):
Screenshot_2020-11-25_23-47-47
Church Latin (i-gnoscitur):
Screenshot_2020-11-25_23-48-01

**Comment by poire-z** [comment]
Thanks for the cultural insights :)
So, no more comments, ready to be merged and bumped?

**Comment by jperon** [comment]
I think it's ready.



Just for kidding, let's get the sample size to 2 (same reference, next page)…

image

But as I said, IMO that's not the question!

**Comment by poire-z** [comment]
Doing the crengine bump into base.

**Comment by Frenzie** [comment]
> Just for kidding, let's get the sample size to 2 (same reference, next page)…

lol, nice

> But as I said, IMO that's not the question!

I tend to expect classical unless it's Catholic in origin.

> most readers of philosophical/theological books from (roughly) S. Augustine to (roughly) the 18th - 19th century would be surprised by mag-nus, while it would be the opposite for "classical" readers or today universities.

I'd expect something more akin to classical in any case.

For the record, I'm not sure that the hyphenation page you linked is actually correct about magnus. The `gn` is the same sound as `ng` in Dutch/English (like in ring). Therefore it's an exception to the general rule which would lead to mag-nus if applied indiscriminately. I.e., cres-cit, Chris-tus, etc. are correct, but not mag-nus since it is normally ma-gnus. (By Dutch spelling and hyphenation rules it should be mang-nus.) The difference with Church Latin is that it's something more like mangus or mang-ngus.

**Pull request merged**